diff --git a/src/cmd.zig b/src/cmd.zig
index c287e64f..2550ec55 100644
--- a/src/cmd.zig
+++ b/src/cmd.zig
@@ -1446,7 +1446,7 @@ fn replayOffline(
         .replay_threads = cfg.replay_threads,
         .disable_consensus = cfg.disable_consensus,
         .voting_enabled = false,
-        .vote_account_address = Pubkey.ZEROES,
+        .vote_account_address = null,
         .stop_at_slot = cfg.stop_at_slot,
     });
     defer replay_service_state.deinit(allocator);
diff --git a/src/replay/freeze.zig b/src/replay/freeze.zig
index ebca08a4..fb5de047 100644
--- a/src/replay/freeze.zig
+++ b/src/replay/freeze.zig
@@ -40,6 +40,8 @@ pub const FreezeParams = struct {
     hash_slot: HashSlotParams,
     finalize_state: FinalizeStateParams,
 
+    stop_at_slot: ?Slot = null,
+
     pub fn init(
         logger: Logger,
         account_store: AccountStore,
@@ -48,6 +50,7 @@ pub const FreezeParams = struct {
         constants: *const SlotConstants,
         slot: Slot,
         blockhash: Hash,
+        stop_at_slot: ?Slot,
     ) FreezeParams {
         return .{
             .logger = logger,
@@ -83,6 +86,7 @@ pub const FreezeParams = struct {
                 .collected_transaction_fees = state.collected_transaction_fees.load(.monotonic),
                 .collected_priority_fees = state.collected_priority_fees.load(.monotonic),
             },
+            .stop_at_slot = stop_at_slot,
         };
     }
 };
@@ -112,7 +116,10 @@ pub fn freezeSlot(allocator: Allocator, params: FreezeParams) !void {
         "froze slot {} with hash {s}",
         .{ params.hash_slot.slot, slot_hash.get().*.?.base58String().slice() },
     );
-
+    try std.io.getStdOut().writer().print("slot={} hash={}\n", .{ params.hash_slot.slot, slot_hash.get().*.? });
+    if (params.stop_at_slot) |stop| if (params.hash_slot.slot >= stop) {
+        std.process.exit(0);
+    };
     // NOTE: agave updates hard_forks and hash_overrides here
 }
 
@@ -443,7 +450,7 @@ test "freezeSlot: trivial e2e merkle hash test" {
 
     try freezeSlot(
         allocator,
-        .init(.FOR_TESTS, account_store, &epoch, &state, &constants, 0, .ZEROES),
+        .init(.FOR_TESTS, account_store, &epoch, &state, &constants, 0, .ZEROES, null),
     );
 
     try std.testing.expectEqual(
@@ -502,7 +509,7 @@ test "freezeSlot: trivial e2e lattice hash test" {
 
         try freezeSlot(
             allocator,
-            .init(.FOR_TESTS, account_store, &epoch, &state, &constants, 0, .ZEROES),
+            .init(.FOR_TESTS, account_store, &epoch, &state, &constants, 0, .ZEROES, null),
         );
 
         try std.testing.expectEqual(
diff --git a/src/replay/service.zig b/src/replay/service.zig
index 09275ccb..18154f6e 100644
--- a/src/replay/service.zig
+++ b/src/replay/service.zig
@@ -524,6 +524,7 @@ fn freezeCompletedSlots(state: *ReplayState, results: []const ReplayResult) !boo
                     slot_info.constants,
                     slot,
                     last_entry_hash,
+                    state.stop_at_slot,
                 ));
                 processed_a_slot = true;
             } else {
diff --git a/src/shred_network/collector/shred_tracker.zig b/src/shred_network/collector/shred_tracker.zig
index 8bd10f7a..44f6795b 100644
--- a/src/shred_network/collector/shred_tracker.zig
+++ b/src/shred_network/collector/shred_tracker.zig
@@ -276,6 +276,10 @@ pub const BasicShredTracker = struct {
         }
         self.current_bottom_slot = @max(self.current_bottom_slot, slot);
         self.metrics.finished_slots_through.max(slot -| 1);
+        self.logger.info().logf(
+            "advanced bottom slot to {}",
+            .{self.current_bottom_slot},
+        );
     }
 
     /// - Record that a slot has been observed.
